sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketsInJoin.scala

```diff
@@ -0,0 +1,116 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.bucketing
+
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.plans.Inner
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This rule injects a hint if one side of two bucketed tables can be coalesced
+ * when the two bucketed tables are inner-joined and they differ in the number of buckets.
+ */
+object CoalesceBucketsInJoin extends Rule[LogicalPlan]  {
+  private def isPlanEligible(plan: LogicalPlan): Boolean = {
+    def forall(plan: LogicalPlan)(p: LogicalPlan => Boolean): Boolean = {
+      p(plan) && plan.children.forall(forall(_)(p))
+    }
+
+    forall(plan) {
+      case _: Filter | _: Project | _: LogicalRelation => true
+      case _ => false
+    }
+  }
+
+  private def getBucketSpec(plan: LogicalPlan): Option[BucketSpec] = {
+    if (isPlanEligible(plan)) {
+      plan.collectFirst {
+        case _ @ LogicalRelation(r: HadoopFsRelation, _, _, _) if r.bucketSpec.nonEmpty =>
+          r.bucketSpec.get
+      }
+    } else {
+      None
+    }
+  }
+
+  private def mayCoalesce(
+      numBuckets1: Int,
+      numBuckets2: Int,
+      maxNumBucketsDiff: Int): Option[Int] = {
+    assert(numBuckets1 != numBuckets2)
+    val (small, large) = (math.min(numBuckets1, numBuckets2), math.max(numBuckets1, numBuckets2))
+    // A bucket can be coalesced only if the bigger number of buckets is divisible by the smaller
+    // number of buckets because bucket id is calculated by modding the total number of buckets.
+    if ((large % small == 0) &&
+      ((large - small) <= maxNumBucketsDiff)) {
+      Some(small)
+    } else {
+      None
+    }
+  }
+
+  private def addCoalesceBuckets(plan: LogicalPlan, numCoalescedBuckets: Int): LogicalPlan = {
+    plan.transformUp {
+      case l @ LogicalRelation(_: HadoopFsRelation, _, _, _) =>
+        CoalesceBuckets(numCoalescedBuckets, l)
+    }
+  }
+
+  object ExtractJoinWithBuckets {
+    def unapply(plan: LogicalPlan): Option[(Join, Int, Int)] = {
+      plan match {
+        case join: Join =>
```

Comment:
This optimization still works well even for full-outer joins? I think this targets at hash-based joins though.

Reply:
It will remove shuffle for full-outer join as well:
With the feature off:
```
scala> t1.join(t2, t1("i") === t2("i"), "full_outer").explain
== Physical Plan ==
SortMergeJoin [i#67], [i#73], FullOuter
:- *(2) Sort [i#67 ASC NULLS FIRST], false, 0
:  +- Exchange hashpartitioning(i#67, 200), true, [id=#418]
:     +- *(1) ColumnarToRow
:        +- FileScan parquet default.t1[i#67,j#68,k#69] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 8 out of 8
+- *(4) Sort [i#73 ASC NULLS FIRST], false, 0
   +- Exchange hashpartitioning(i#73, 200), true, [id=#425]
      +- *(3) ColumnarToRow
         +- FileScan parquet default.t2[i#73,j#74,k#75] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 4 out of 4
```
With the feature on:
```
scala> t1.join(t2, t1("i") === t2("i"), "full_outer").explain
== Physical Plan ==
SortMergeJoin [i#67], [i#73], FullOuter
:- *(1) Sort [i#67 ASC NULLS FIRST], false, 0
:  +- *(1) ColumnarToRow
:     +- FileScan parquet default.t1[i#67,j#68,k#69] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 8 out of 8
+- *(2) Sort [i#73 ASC NULLS FIRST], false, 0
   +- *(2) ColumnarToRow
      +- FileScan parquet default.t2[i#73,j#74,k#75] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex[], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i:int,j:int,k:string>, SelectedBucketsCount: 4 out of 4
```
Did I miss something?

Reply:
Ah, I see. I misunderstood something. One more comment; we need to check here if join is a equi-join or not?
```
case join: Join(_, _, _, Some(equality condition), _) =>
```

Reply:
Yes, good idea. I changed it to apply coalescing only for equi-joins using `ExtractEquiJoinKeys`. Please let me know what you think.